### PR TITLE
add support for provisioned concurrency

### PR DESCRIPTION
**Pull request by kamaz**
## Proposed changes

Not long time ago AWS introduced ability to create provisioned concurrency which was recently added to `serverless` framework. At the moment plugin is not compatible with the feature. When used it causes a service to stop responding after the switch because permissions for the lambda are not assigned for API Gateway. 

Additionally added autoscaling option for lambdas.

Would be good to hear your feedback on the change

## Types of changes

What types of changes does your code introduce to the plugin?
_Put an `x` in the boxes that apply_

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

## Further comments

No additional comments 😄

### CODE REVIEW

**Comment by davidgf** [comment]
Hey @kamaz, first of all, thank you very much for you PR! I would need you to change the following:

1. Resolve the conflicts, as I've merged a previous PR
2. Revert the `package.json` version change. With many people contributing, a new PR doesn't necessarily mean new version
3. Add the needed config to test the new features in the [example service](https://github.com/davidgf/serverless-plugin-canary-deployments/tree/master/example)

Thanks!

File: example/package.json [hunk 1]
```diff
@@ -8,14 +8,14 @@
     "uuid": "^3.3.2"
   },
   "devDependencies": {
-    "serverless": "^1.47.0",
+    "serverless": "^1.59.0",
```
Comment (kamaz) [inline example/package.json]: for provisioning to work, this is the minimum version required

File: example/package.json [hunk 2]
```diff
@@ -8,14 +8,14 @@
...
     "serverless-plugin-aws-alerts": "^1.2.4",
-    "serverless-plugin-canary-deployments": "^0.4.2"
+    "serverless-plugin-canary-deployments": "^0.4.8"
   },
   "scripts": {
     "test": "echo \"Error: no test specified\" && exit 1",
-    "package": "npm un --no-save serverless-plugin-canary-deployments && npm pack ../ && npm i --no-save serverless-plugin-canary-deployments-0.4.7.tgz && sls package -s dev",
-    "deploy": "npm un --no-save serverless-plugin-canary-deployments && npm pack ../ && npm i --no-save serverless-plugin-canary-deployments-0.4.7.tgz && sls deploy -s dev",
+    "package": "npm un --no-save serverless-plugin-canary-deployments && npm pack ../ && npm i --no-save serverless-plugin-canary-deployments-0.4.8.tgz && sls package -s dev",
```
Comment (kamaz) [inline example/package.json]: up the version to `0.4.8` as it was failing to package and deploy

File: example/serverless.yml [hunk 3]
```diff
@@ -29,9 +29,10 @@ functions:
       alias: Live
   hello:
     handler: handler.hello
+    provisionedConcurrency: 1
     events:
       - websocket:
-          route: $default
+          route: $connect
```
Comment (kamaz) [inline example/serverless.yml]: I was getting following error: 
```
An error occurred: SdefaultWebsocketsRoute - Currently, authorization is restricted to the $connect route only (Service: AmazonApiGatewayV2; Status Code: 400; Error Code: BadRequestException; Request ID: 6933e404-12a5-4288-a072-217a7dfab05d).
```

File: example/serverless.yml [hunk 4]
```diff
@@ -50,7 +51,7 @@ functions:
       - schedule:
           rate: rate(1 minute)
           enabled: false
-      - cloudwatchLog: "/aws/lambda/sls-wshop-dev-hello"
+      # - cloudwatchLog: '/aws/lambda/sls-wshop-dev-hello'
```
Comment (kamaz) [inline example/serverless.yml]: wasn't sure about this. I was testings using this application and that cloudwatch log does not exist, as one would expect 

I've commented out

**Comment by kamaz** [comment]
Hey @davidgf  

Sorry, it took a little bit longer than I've expected. I was snowed under with work this week but finally, I got around addressing the issue. 

Thanks for the plugin it works really well and like how it is done. 

Also added screenshot after deploying example service below with provisioning enabled:

Screenshot 2020-02-07 at 20 34 59

**Comment by jogura** [comment]
Is there any update on this fix? 

I also appreciate this plugin and everyone's hard work on it.

**Comment by davidgf** [comment]
Hi @kamaz! First of all, thank you very much for all the effort. I need clarification on what the `autoscaling` feature does in the context of Lambda, and why is that something that needs to be included in a plugin that manages canary deployments. Also, there are many many changes that are simply aesthetic (splitting a line of code into several ones, etc.) and don't change the business logic at all, which is making the review process much harder.

**Comment by kamaz** [comment]
Hi @davidgf 

No worries. 

This [link](https://aws.amazon.com/blogs/aws/new-provisioned-concurrency-for-lambda-functions/) may expand on purpose of `autoscaling`. Because provisioned concurrency and `autoscalling` are related in this context and how we use it I though I will add it to the plugin. 

If you would prefer for me to remove it. I'm happy to do it. 

I'm sorry for a little bit mess there when it comes to formatting. It must've been my VSCode. Happy to reverse them if that cause problems?

**Comment by asheliahut** [comment]
I really want this too, but I tend to agree with @davidgf the autoscaling portion probably is best if went with serverless itself not in a plugin. If you read through it. The provisioned concurrency doesn't really need to be set by this plugin.

I still haven't gotten around to doing it, but the best solution to all of this is to fix it at the serverless level by allowing a custom name on the provisioned alias. At least that is my opinion.

**Comment by davidgf** [comment]
Hey Kamaz, I'm very sorry for the late reply, due to the current situation I couldn't spend any time on the plugin. I'm going to close the PR because this error was introduced in the Serverless version `v1.60.2`, but it was already fixed in `v1.61.0`. You could argue that the plugin wouldn't work with functions that have provisioned concurrency setup, and you'd be right, but I think that this is part of a much broader discussion. Basically, the Serverless core framework doesn't support aliases, it just maps all the event sources to the Lambda `$LATEST` qualifier. However, there are some features that cannot be used with `$LATEST` and need an alias, such as provisioned concurrency or weighted aliases (the feature that this plugin uses to handle canary deployments). In order to solve that, the Serverless core creates now an alias only for those functions that have provisioned concurrency setup. However, this alias cannot be customized, and its name is a hardcoded string ([`provisioned`](https://github.com/serverless/serverless/blob/7479a9ae82b44fb06de3ab84094b18e8f72affc4/lib/plugins/aws/lib/naming.js#L164)). In my humble opinion, this is not a long term solution, because it pretty much conflicts with this or any other plugins that need to leverage and manipulate aliases. Ideally, aliases should be centrally managed by the core, rather than only creating an alias for a specific use case. I'm trying to reach out to the framework maintainers to hear whether they're open to these kind of ideas, and I'm sure they'll hear them, because just by doing a quick search on their repo issues, there are a few of them related to this topic (such as https://github.com/serverless/serverless/issues/7209 or https://github.com/serverless/serverless/issues/6849)

**Comment by jweyrich** [comment]
@davidgf I agree with your arguments above. It should be something supported by the Serverless framework itself. Are you aware of any Serverless v3 plugin that support aliases? We'd like to use the alias feature, but couldn't find any maintained plugin.

**Pull request closed without merging**